File: Data Structures & Algorithms/combination-target-sum-ii/submission-0.java
class Solution {
    private List<List<Integer>> res = new ArrayList<>();
    private List<Integer> combi = new ArrayList<>();
    private void backtrack(int[] nums, int begin, int target) {
        if(target == 0) {
            res.add(new ArrayList<>(combi));
            return;
        }
        for(int i = begin; i >= 0; i--) {
            if(i < begin && nums[i] == nums[i + 1]) {
                continue;
            }
            combi.add(nums[i]);
            backtrack(nums, i-1, target - nums[i]);
            combi.remove(combi.size() - 1);
        }
    }

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        if(candidates.length > 1){
            Arrays.sort(candidates);
        }
        backtrack(candidates, candidates.length - 1, target);
        return res;
    }
}

File: Data Structures & Algorithms/binary-search/submission-0.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int i = 0, j = nums.size() - 1;
        while(i <= j) {
            int m = i + (j-i)/2;
            if(nums[m] == target) return m;
            if(nums[m] < target) i=m+1;
            else j = m-1;
        }
        return -1;
    }
};
